### ISSUE TITLE
TFLite quant model run on the android phone have no speed up than float model

### ISSUE THREAD

**Issue by hahadashi** [tensorflow/tensorflow#40914]
i train a CNN + DNN model by Quantization-aware training way, get two tflite model (quant 945k and float 3.7M, same structure). but in the phone there is no speed up.

**Comment by amahendrakar** [comment]
@hahadashi,
In order to expedite the trouble-shooting process, could you please provide the complete code to reproduce the issue reported here and also the TensorFlow version you are using. Thanks!

**Comment by hahadashi** [comment]
@amahendrakar `import sys
import tensorflow as tf
import tensorflow_model_optimization as tfmot
import numpy as np
from absl import logging
import time
import os

def save_to_float_tflite(model, step):
    converter = tf.lite.TFLiteConverter.from_keras_model(model)
     #converter.optimizations = [tf.lite.Optimize.DEFAULT]
    tflite_model = converter.convert()
    open("tflite_float/model_"+str(step)+".tflite", "wb").write(tflite_model)

def save_to_quant_tflite(model, step):
    #converter = tf.lite.TFLiteConverter.from_concrete_functions([concrete_func])
    #converter = tf.lite.TFLiteConverter.from_saved_model("saved_model/1")
    converter = tf.lite.TFLiteConverter.from_keras_model(model)
    converter.optimizations = [tf.lite.Optimize.DEFAULT]
    #converter.experimental_new_converter = True
    tflite_model = converter.convert()
    open("tflite_quant/model_"+str(step)+".tflite", "wb").write(tflite_model)

class Trainer:
    def __init__(self):
        gpus = tf.config.experimental.list_physical_devices("GPU")
        for gpu in gpus:
            tf.config.experimental.set_memory_growth(gpu, True)
        if gpus:
            tf.config.experimental.set_visible_devices(gpus[0], "GPU")
        self.optimizer = tf.keras.optimizers.Adam(learning_rate=0.001)
    def get_loss(self, logits, input_labels, training=True):
        loss_func = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
        acc = self.get_train_acc(logits, input_labels)
        return loss_func(input_labels, logits), acc
    
    def get_train_acc(self, logits, input_labels):
        pred = tf.math.argmax(logits, axis=1)
        train_acc = tf.reduce_mean(tf.cast(tf.math.equal(pred, input_labels), dtype=tf.float32))
        return train_acc
     
    def clip_by_norm(self, grads, norm):
        """ clip norm using tf.clip_by_norm """
        if norm <= 0:
            return grads
        grads = [
            None if gradient is None else tf.clip_by_norm(gradient, norm)
            for gradient in grads
        ]
        return grads

    def train_step(self, model, input_feats, input_labels):
        """ train the model 1 step """
        with tf.GradientTape() as tape:
            logits = model(input_feats)
            loss, train_acc = self.get_loss(logits, input_labels, training=True)
        # Horovod: add Horovod Distributed GradientTape.
        grads = tape.gradient(loss, model.trainable_variables)
        grads = self.clip_by_norm(grads, cfg.clip_norm)
        self.optimizer.apply_gradients(zip(grads, model.trainable_variables))
        return loss, train_acc

    def train(self):
        layers = tf.keras.layers
        input_features = layers.Input(shape=(21, 120), dtype=tf.float32, name="input")
        inner = layers.Reshape(target_shape=(21, 120, 1))(input_features)
        inner = layers.Conv2D(
            filters=num_filters1,
            kernel_size=(15, 4),
            strides=(2, 2),
            padding="same",
            use_bias=True,
            data_format="channels_last",
        )(inner)
        inner = layers.BatchNormalization()(inner)
        inner = layers.MaxPool2D(pool_size=(1, 3), strides=(1, 3), padding='SAME')(inner) 
        #inner = tf.nn.relu(inner)
        inner = layers.ReLU()(inner)
        inner = layers.Conv2D(
            filters=num_filters2,
            kernel_size=(4, 3),
            strides=(1, 2),
            padding="same",
            use_bias=True,
            data_format="channels_last",
        )(inner)
        inner = layers.BatchNormalization()(inner)
        #inner = tf.nn.relu(inner)
        inner = layers.ReLU()(inner)
        inner = layers.Flatten()(inner)
        inner = layers.Dense(258, activation=tf.nn.relu)(inner)
        #inner = layers.BatchNormalization()(inner)
        #inner = layers.ReLU()(inner) 
        inner = layers.Dense(258, activation=tf.nn.relu)(inner)
        #inner = layers.BatchNormalization()(inner)
        #inner = layers.ReLU()(inner)
        inner = layers.Dense(10, name="output")(inner)
        model = tf.keras.Model(inputs=input_features, outputs=inner, name="kws")
        quantize_model = tfmot.quantization.keras.quantize_model
        #model = quantize_model(model_1)
        print(model.summary())
        model_quant = quantize_model(model)
        print(model_quant.summary()) 
        epoch = global_step = 0
        while epoch < cfg.MAX_EPOCH:
            logging.info(">>>>> start training in epoch %d" % epoch)
            while True:
                # train 1 step
                batch_feats, batch_labels, epoch = # push some data
                if global_step >= 2000000:
                    loss, train_acc = self.train_step(model_quant, batch_feats, batch_labels) 
                    if global_step % cfg.save_model_iter == 0:
                        inner = layers.Softmax()(model.outputs)
                        model1 = tf.keras.Model(inputs=input_features, outputs=inner, name="kws_quant")
                        save_to_quant_tflite(quantize_model(model1), global_step)
                else:
                    loss, train_acc = self.train_step(model, batch_feats, batch_labels)
                    if global_step == 1999999 or global_step == 2000000:
                        save_to_float_tflite(model, global_step) 
                 
                global_step += 1

if __name__ == "__main__":
    logging.set_verbosity(logging.INFO)
    tf.random.set_seed(1)
    trainer = Trainer()
    trainer.train()
`

it is my demo, i test two model (float and quant), the speed is same. hope you give some advice.

**Comment by hahadashi** [comment]
@amahendrakar hi, can you find some problems from my model.

**Comment by jdduke** [comment]
Which TFLite runtime libraries are you using? Did you built yourselves or are you using the prebuilts? If the prebuilts, which version? If you built yourselves, what build command did you use? And what phone are you testing on? How did you measure performance?

Quantization should generally be faster than CPU, but it's not always the case, depending on the model and the device.

**Comment by hahadashi** [comment]
> Did you built yourselves or are you using the prebuilts?

thxs your reply. i clone tf2.0 and build tflite so libraries(bazel build --config=opt -c opt --copt=-mavx2 --cxxopt=--std=c++11 //tensorflow/lite:libtensorflowlite_c.so). what`s mean of you prebuilt?  device : Qualcomm SoCs Snapdragon 660. i measure the rtf of model.

**Comment by jdduke** [comment]
@srjoglekar246 can you help diagnose?

**Comment by srjoglekar246** [comment]
@hahadashi is this is still relevant, can you provide the sample float & quantized models you are testing? Its difficult to judge based on your code why the model performance might be off.

**Comment by tilakrayal** [comment]
@jdduke ,
Can you please provide the details mentioned above.It helps to debug the issue.Thanks!
